contributed_definitions/nyaml/NXcalibration.yaml

```diff
@@ -27,14 +27,25 @@ NXcalibration(NXobject):
     dimensions:
       rank: 1
       dim: [[1, ncoeff]]
+  reference(NXidentifier):
+    doc: |
+      A reference to a calibration on which this calibration is based.
```

Comment:
I don't understand this docstring. Maybe i can't remember what was decided here, but I thought this was meant to be used for indicating that the calibration was done externally and that information about that calibration can be found somewhere else?

Reply:
It could also be an entry in an RDM system, not just a file, right? But this is probably nitpicking.

Reply:
if file my thought was use NXserialized

Reply:
No, we are talking about a doi, which in my view doesn't have to be an actual file as also @lukaspie mentioned. So we should get rid of this wording.

Like so?
```
A digital persistent identifier (e.g., a doi) referring to an object containing the calibration applied here.
The axis values might also be copied/linked here for reference.
```

Reply:
NXidentifier was designed with the aim in mind to avoid having to detail in the docstring which type of service is used as there is more to the stating of we have an "identifier or reference for something here"

I thought the following, if the calibration is in a file and you would like to refer to this file - but do not like to copy over the content of that file into NeXus (for whatever reason) NXserialized comes to rescue

But if you would like to say hey that calibration is detailed in some literature reference with that particular doi
use either:
NXcite
or
reference(NXidentifier)
service=doi
identifier=your value 10.7232.... whatever
is_persistent=TRUE

Reply:
Ok I see. So I think NXcite is better suited here. I changed the docstring and base class accordingly

Reply:
Do ISO standards have DOIs? How would I reference [this](https://www.iso.org/standard/55796.html)?

Reply:
I think we are thinking about different things for this field. My view is that it should be a reference to a file which contains an actual measurement carried out with which the data was calibrated, e.g., a standard reference sample and **not** a method reference for how a calibration has been done.

The latter one is important, too. So we probably should add both and distinguish them clearly in the docstrings. What do you think @lukaspie?

We could add a `calibration_method` field for this.
Additionally, with NXcite you can also just refer to an url: https://manual.nexusformat.org/classes/base_classes/NXcite.html

Reply:
> Do ISO standards have DOIs? How would I reference [this](https://www.iso.org/standard/55796.html)?

"Standards that the ISO and IEC jointly develop are identified by the prefix "ISO/IEC." An example of this approach is ISO/IEC 27001:2013. It specifies requirements for setting up and running an information security management system."
Adding "iso" as a service for NXidentifier and then use the "27001:2013" as the value, is_persistent=? mmhh, debateable but will you get anything more persistent than that?

Reply:
I guess NXcite is not flexible enough for what we want to do, so maybe NXidentifier is better? But then if we restrict the service list in NXidentifier , what do we do if the service is not in `[doi, urn, hdl, purl, orcid, iso]`?

I am not sure about `calibration_method`, I would probably expect a free-text description of the actual process. Maybe we can move `description(NX_CHAR)` right before this so it is clear that this free-form text should go to this field.

How about this?
```yaml
NXcalibration.yaml
(NXcalibration):
  description(NX_CHAR):
    doc: |
      A description of the procedures employed.
  calibration_method(NXidentifier):
    doc: |
      A digital persistent identifier (e.g., a doi) referring to the description of the calibration
      method applied here.
  reference_serialized(NXserialized):           
    doc: |
     Metadata to a set of pieces of information describing the calibration event.
      
     This metadata can be a documentation of the source (file) or database (entry) from which pieces
     of information have been extracted for consumption in e.g. a research data management system (RDMS).
     It is also possible to include the actual file by using the `file` field.
 
     The axis values might also be copied/linked here for reference.

NXserialized.yaml
category: base
doc: |
  Metadata to a set of pieces of information of a resource that has been serialized.
  
  A typical use case is the documentation of the source (file) or database (entry)
  from which pieces of information have been extracted for consumption in e.g. a
  research data management system (RDMS). This may be for reasons of enabling
  services such as providing access to normalized information for which reading
  again from the resource may not be desired, possibe, or feasible.
  
  Possible reasons could be the extraction of specific information for caching,
  performance reasons, or re-evaluate given pieces of information based on other
  views and interaction patterns with the data where information has been formatted
  differently by tools than how these pieces of information were originally
  serialized.
type: group
NXserialized(NXobject):
  type(NX_CHAR):
    doc: |
      Answers into what resource the information was serialized.
    enumeration: [file, database]
  path(NX_CHAR):
    doc: |
      Path to the resource.
      
      E.g. the name of a file or its absolute or relative path, or the
      identifier to a resource in another database.
  checksum(NX_CHAR):
    doc: |
      Value of the checksum obtain when running algorithm on the resource.
  algorithm(NX_CHAR):
    doc: |
      Name of the algorithm whereby the checksum was computed.
    enumeration: [md5, sha256]
  file(NXnote):
    Extracted file containing the serialized information.
     
NXidentifier.yaml
category: base
doc: |
  Resource identifier.
type: group
NXidentifier(NXobject):
  service(NX_CHAR):
    doc: |
      The service/type whereby the identifier is resolvable.
    enumeration: [doi, urn, hdl, purl, orcid, iso] # should probably be extended
  identifier(NX_CHAR):
    doc: |
      Value/code of the identifier.
  is_persistent(NX_BOOLEAN):
    doc: |
      Answers, based on type, if the identifier, is a persistent one or not.
```

If we add `file` in NXserialized, we can also extract the whole file and store it in NeXus if needed. But we can also remove this if we don't want to have any non-NeXus information here.

Reply:
> I guess NXcite is not flexible enough for what we want to do, so maybe NXidentifier is better? But then if we restrict the service list in NXidentifier , what do we do if the service is not in `[doi, urn, hdl, purl, orcid, iso]`?

We could add `url` to the enumeration as well. So you can select a certain service but also fall back to a simple url for other resources. All of the present ones could also resolve to a url, right? So this would fit.

> I am not sure about `calibration_method`, I would probably expect a free-text description of the actual process. Maybe we can move `description(NX_CHAR)` right before this so it is clear that this free-form text should go to this field.

Yes, you are right. Probably it's better to have the NXidentifier and add a free text description if necessary.
> 
> How about this?
> 
> ```yaml
> NXcalibration.yaml
> (NXcalibration):
>   description(NX_CHAR):
>     doc: |
>       A description of the procedures employed.
>   calibration_method(NXidentifier):
>     doc: |
>       A digital persistent identifier (e.g., a doi) referring to the description of the calibration
>       method applied here.
>   reference_serialized(NXserialized):           
>     doc: |
>      Metadata to a set of pieces of information describing the calibration event.
>       
>      This metadata can be a documentation of the source (file) or database (entry) from which pieces
>      of information have been extracted for consumption in e.g. a research data management system (RDMS).
>      It is also possible to include the actual file by using the `file` field.
>  
>      The axis values might also be copied/linked here for reference.
> 
> NXserialized.yaml
> category: base
> doc: |
>   Metadata to a set of pieces of information of a resource that has been serialized.
>   
>   A typical use case is the documentation of the source (file) or database (entry)
>   from which pieces of information have been extracted for consumption in e.g. a
>   research data management system (RDMS). This may be for reasons of enabling
>   services such as providing access to normalized information for which reading
>   again from the resource may not be desired, possibe, or feasible.
>   
>   Possible reasons could be the extraction of specific information for caching,
>   performance reasons, or re-evaluate given pieces of information based on other
>   views and interaction patterns with the data where information has been formatted
>   differently by tools than how these pieces of information were originally
>   serialized.
> type: group
> NXserialized(NXobject):
>   type(NX_CHAR):
>     doc: |
>       Answers into what resource the information was serialized.
>     enumeration: [file, database]
>   path(NX_CHAR):
>     doc: |
>       Path to the resource.
>       
>       E.g. the name of a file or its absolute or relative path, or the
>       identifier to a resource in another database.
>   checksum(NX_CHAR):
>     doc: |
>       Value of the checksum obtain when running algorithm on the resource.
>   algorithm(NX_CHAR):
>     doc: |
>       Name of the algorithm whereby the checksum was computed.
>     enumeration: [md5, sha256]
>   file(NXnote):
>     Extracted file containing the serialized information.
>      
> NXidentifier.yaml
> category: base
> doc: |
>   Resource identifier.
> type: group
> NXidentifier(NXobject):
>   service(NX_CHAR):
>     doc: |
>       The service/type whereby the identifier is resolvable.
>     enumeration: [doi, urn, hdl, purl, orcid, iso] # should probably be extended
>   identifier(NX_CHAR):
>     doc: |
>       Value/code of the identifier.
>   is_persistent(NX_BOOLEAN):
>     doc: |
>       Answers, based on type, if the identifier, is a persistent one or not.
> ```
> 
> If we add `file` in NXserialized, we can also extract the whole file and store it in NeXus if needed. But we can also remove this if we don't want to have any non-NeXus information here.

I like to possibility of storing the whole file here. But there is still the problem if we have a binary file. Can we store this as well in an `NXnote`. Other than that I think it's good to use NXidentifier and NXserialized.

Reply:
I added `calibration_reference(NXidentifier)` and `calibration_object(NXserialized)` now to either refer to a published calibration or just a single file/entry somewhere. Ideally, I would like to have something like `calibration_reference(NXidentifier | NXserialized)` here, where we could select which one of the two we specify (there is an [choice](https://manual.nexusformat.org/defs_intro.html#choice) in NeXus but I think we don't support this, so I think we should just go with the duplicating pattern here).

Reply:
@domna indeed url can be added. Intentionally I did not want to make the NXidentifier matching only the concept of a persistent such identifier.

The reason why I went for enumeration and not a free-text is that with a free text we have no chance for ensuring that typos are catched and thus boiler plate code in the RDMS or elsewhere is needed to harmonize the name of the service
but I do agree I felt also split in between the chairs when I wrote enumeration as I for sure do not know all possible values for the enum but maybe here is what we can do best as a common effort and group

Reply:
@domna that NXidentifier | NXserialized though I like and supporting that would indeed solve also the case that I mentioned in the standup 
 triangle in 2d box(NXcg_parallelogram_set) while triangle in 3d box(NXcg_hexahedron_set)
 but ideally I would like to have
 box(NXcg_parallelogram_set | NXcg_hexahedron_set) as NXcg_triangle_set has a symbol dimensionality with enum values 2 and 3 to avoid having to create two base classes for a thing that remains a set of triangles

Reply:
I think having a url should be fine. Most services resolve as a url anyways so it's broad enough for this case here. If someone wants to specify a particular one this is possible as well. 

Another thought out of the scope here: For me there is still the question if we should not simply use and generalise NXcite for this and add additional fields such as orcid, purl, persistent, etc. I have the feeling NXidentifier is just using another design pattern for referring/citing an external (persistent) resource. We could also let NXcite extend NXidentifier (or the other way around). If we want something like this we should add this to the backlog.

Reply:
> @domna that (NXidentifier | NXserialized) thought I like and supporting that would indeed solve also the case that I mentioned in the standup triangle in 2d box(NXcg_parallelogram_set) while triangle in 3d box(NXcg_hexahedron_set) but ideally I would like to have box(NXcg_parallelogram_set | NXcg_hexahedron_set) as NXcg_triangle_set has a symbol dimensionality with enum values 2 and 3 to avoid having to create two base classes for a thing that remains a set of triangles But indeed that thought is not implemented currently

Yes, we should probably add this to backlog as well. But I think this is related to base class inheritance as well and might be solved by/with it (https://github.com/FAIRmat-NFDI/pynxtools/issues/166)

Reply:
In general, like the idea of supporting choice (e.g., by the (NXidentifier | NXserialized) logic). This should be added to the backlog. It could possibly be solved by base class inheritance, by maybe there are also cases where you really have a choice between two quite distinct base classes.

In this case, however, why now have both `calibration_reference(NXidentifier)` _and_ `calibration_object(NXserialized)`. `calibration_reference(NXidentifier)` to identify a literature (or similar) reference where the method is described and `calibration_object(NXserialized)` to actually link to the instance of this calibration, e.g. a file.

> I like to possibility of storing the whole file here. But there is still the problem if we have a binary file. Can we store this as well in an NXnote.

NXnote has a 'data: (optional) [NX_BINARY]' field, so binary files could be supported. Maybe we can indicate in the doc string that we prefer not to have binary files, but it is technically possible.

> Another thought out of the scope here: For me there is still the question if we should not simply use and generalise NXcite for this and add additional fields such as orcid, purl, persistent, etc. I have the feeling NXidentifier is just using another design pattern for referring/citing an external (persistent) resource. We could also let NXcite extend NXidentifier (or the other way around). If we want something like this we should add this to the backlog.

Yeah, I agree that NXidentifier and NXcite feel very similar. NXcite should extend NXidentifier, I guess.

Reply:
Or maybe I misunderstood, `calibration_reference(NXidentifier)` is not supposed to be the same as `calibration_method(NXidentifier)`?

Reply:
> NXnote has a 'data: (optional) [NX_BINARY]' field, so binary files could be supported. Maybe we can indicate in the doc string that we prefer not to have binary files, but it is technically possible.

Yes we should discourage this but we should have the possibility, because researchers might not have anything else than a binary file.

> > Another thought out of the scope here: For me there is still the question if we should not simply use and generalise NXcite for this and add additional fields such as orcid, purl, persistent, etc. I have the feeling NXidentifier is just using another design pattern for referring/citing an external (persistent) resource. We could also let NXcite extend NXidentifier (or the other way around). If we want something like this we should add this to the backlog.
> 
> Yeah, I agree that NXidentifier and NXcite feel very similar. NXcite should extend NXidentifier, I guess.

👍️

> Or maybe I misunderstood, `calibration_reference(NXidentifier)` is not supposed to be the same as `calibration_method(NXidentifier)`?

IMO they should not be the same. We have three fields now, which I would describe as:

- calibration_method: A reference to a detailed description of a calibration method but no actual calibration data (i.e., ISO standard)
- calibration_reference: A reference to a publicly available calibration standard used for this instrument, e.g., a measurement of a known standard containing calibration information.
- calibration_object: A file serialisation of a calibration which may not be publicly available (externally from the nexus file). For me the main use case for this is to include instrument calibration in a certain vendor format, which can be very useful for internal use but might not be too useful for someone using a different instrument.

Reply:
Ok, sounds good, I was just a bit confused with having two NXidentifiers. Mabye we can extend the existing docstrings with the explanations given here.

Reply:
```yaml
calibration_method(NXidentifier):
    doc: |
      A digital persistent identifier (e.g., doi, ISO standard) referring to a detailed description of a 
      calibration method but no actual calibration data.
  calibration_reference(NXidentifier):
    doc: |
      A digital persistent identifier (e.g., a doi) referring to a 
      publicly available calibration measurement used for this instrument
      , e.g., a measurement of a known standard containing calibration information.
      The axis values may be copied or linked in the appropriate NXcalibration fields for reference.
  calibration_object(NXserialized):
    doc: |
      A file serialisation of a calibration which may not be publicly available (externally from the nexus file).
      
      This metadata can be a documentation of the source (file) or database (entry) from which pieces
      of information have been extracted for consumption in e.g. a research data management system (RDMS).
      It is also possible to include the actual file by using the `file` field.
  
      The axis values may be copied or linked in the appropriate NXcalibration fields for reference.
```
Good?

Reply:
Thanks for refining the docstrings, from my POV we can merge.

Reply:
I second